analyses/rna-seq-protocol-dge/01-protocol-dge-seg.R

```diff
@@ -0,0 +1,323 @@
+
+if (identical(norm_method, 'tmm')) {
+    counts_object <- calcNormFactors(counts_object, method="TMM")
+    norm_cnt_mat <- cpm(counts_object)
+
+    counts_object <- counts_object[hkgenes, ]
+    norm_cnt_mat <- norm_cnt_mat[hkgenes, ]
+    round_norm_cnt_mat <- round(norm_cnt_mat)
+
+    plot_nm_str <- 'TMM'
+} else if (identical(norm_method, 'uqpgq2')) {
+    counts_object <- DGEList(counts = uq.pgQ2(X = counts_object$counts),
+                             group = group)
+    counts_object <- counts_object[hkgenes, ]
+
+    norm_cnt_mat <- counts_object$counts
+    round_norm_cnt_mat <- round(norm_cnt_mat)
+    counts_object <- calcNormFactors(counts_object, method="none")
+
+    plot_nm_str <- 'UQ-pgQ2'
+}
+### Estimate dispersion
+counts_object <- estimateDisp(
+    counts_object, design=design, prior.df=NULL, trend.method='locfit',
+    tagwise=TRUE)
+png(file.path(plot_outdir, 'estimated_dispersions.png'))
+plotBCV(counts_object)
+n_null_dev <- dev.off()
+#------------ Run LRT --------------------
+print(paste0('Run differential gene expression LRT on',
+             ' poly-A vs stranded RNA-seq...'))
+glm_fit <- glmFit(counts_object, design)
+lrt <- glmLRT(glm_fit)
+
+# Save LRT p-value table and histogram
+lrt_top_tags <- topTags(lrt, adjust.method = "BH", n = Inf)
+lrt_out_df <- lrt_top_tags$table
+stopifnot(identical(colnames(lrt_out_df),
+                    c("logFC", "logCPM", "LR", "PValue", "FDR")))
+colnames(lrt_out_df) <- c("stranded_over_polya_logFC",
+                          "average_logCPM", "LR", "PValue", "BH FDR")
+write.csv(lrt_out_df, file=file.path(table_outdir,
+                                     'stranded_vs_polya_dge_lrt_res.csv'))
+
+# plot and save p-value histogram
+p <- ggplot(lrt_top_tags$table, aes(x=PValue)) +
+    geom_histogram(binwidth = 0.05, center = 0.025) +
+    theme_classic() +
+    scale_x_continuous(expand = expand_scale(mult = c(0, 0.02))) +
+    scale_y_continuous(expand = expand_scale(mult = c(0, 0.05))) +
+    xlab(paste0('stranded vs poly-A RNA-seq DGE ', plot_nm_str,
+                ' LRT p-value')) +
+    ylab('Gene count') +
+    ggtitle(paste0('Histogram of stranded vs poly-A RNA-seq\n',
+                   'differential gene expression ', plot_nm_str,
+                   ' normalized\n',
+                   'LRT p-values\n',
+                   sum(lrt_top_tags$table$PValue < 0.05),
+                   ' genes have p-value < 0.05\n',
+                   sum(lrt_top_tags$table$PValue >= 0.05),
+                   ' genes have p-value >= 0.05\n',
+                   sum(lrt_top_tags$table$FDR < 0.05),
+                   ' genes have BH FDR < 0.05\n',
+                   sum(lrt_top_tags$table$FDR >= 0.05),
+                   ' genes have BH FDR >= 0.05')) +
+    theme(text = element_text(size=15))
+
+ggsave(file.path(plot_outdir, 'stranded_vs_polya_dge_lrt_pvals_histogram.png'),
+       dpi = 300, plot = p, width = 8, height = 7)
```

Comment:
It seems to me that similar code is used for each test exactTest, glmLRT and glmQLFTest, could we add a function instead?

Reply:
Thank you for the suggestion. 

I have considered adding a function for the similar tabulating and plotting code, but I think it is more maintainable to have separate tabulating and plotting code for each test. When making changes to existing tests or adding new tests in the future, the intermediate results need to be checked carefully, and test-specific code may be necessary. For example, the column names of DESeq2 result table are different from edgeR, and the following specific code is added to handle the NAs in p-values. These specific differences may be overlooked when using a function that takes in the test result and various plotting labels. In addition, I think the tabulating and plotting procedures are simple enough for reuse in other scenarios.

```R
# plot and save p-value histogram
# some DESeq2 p-values are NAs due to all-0 normalized read counts
deseq2_plot_df <- deseq2_res_df[, c('pvalue', 'padj')]
deseq2_plot_df[is.na(deseq2_plot_df)] <- 1
```

Reply:
Oh I suggested it just for the edgeR tests since they seem to be the exact same code.

Reply:
It's totally not necessary (just thought of adding it as a suggestion) and I can approve the PR without it too.

Reply:
Sorry that I misunderstood your suggestion. I thought all tests were concerned.

Thank you for your understanding. I think the code might be good as is, as it is simple enough to read without any function. In the future updates, I will refactor the repeated tabulating and plotting procedures, if they become more complex.